vendor/github.com/urfave/cli/help_minimal.go

```diff
@@ -0,0 +1,399 @@
+			if strings.HasPrefix(lastArg, "--") && count == 1 {
+				continue
+			}
+			// match if last argument matches this flag and it is not repeated
+			if strings.HasPrefix(name, cur) && cur != name && !cliArgContains(name) {
+				flagCompletion := fmt.Sprintf("%s%s", strings.Repeat("-", count), name)
+				_, _ = fmt.Fprintln(writer, flagCompletion)
+			}
+		}
+	}
+}
+
+func DefaultCompleteWithFlags(cmd *Command) func(c *Context) {
+	return func(c *Context) {
+		if len(os.Args) > 2 {
+			lastArg := os.Args[len(os.Args)-2]
+			if strings.HasPrefix(lastArg, "-") {
+				printFlagSuggestions(lastArg, c.App.Flags, c.App.Writer)
+				if cmd != nil {
+					printFlagSuggestions(lastArg, cmd.Flags, c.App.Writer)
+				}
+				return
+			}
+		}
+		if cmd != nil {
+			printCommandSuggestions(cmd.Subcommands, c.App.Writer)
+		} else {
+			printCommandSuggestions(c.App.Commands, c.App.Writer)
+		}
+	}
+}
+
+// ShowCommandHelpAndExit - exits with code after showing help
+func ShowCommandHelpAndExit(c *Context, command string, code int) {
+	_ = ShowCommandHelp(c, command)
+	os.Exit(code)
+}
+
+// ShowCommandHelp prints help for the given command
+func ShowCommandHelp(ctx *Context, command string) error {
+	// show the subcommand help for a command with subcommands
+	if command == "" {
+		HelpPrinter(ctx.App.Writer, SubcommandHelpTemplate, ctx.App)
+		return nil
+	}
+
+	for _, c := range ctx.App.Commands {
+		if c.HasName(command) {
+			templ := c.CustomHelpTemplate
+			if templ == "" {
+				templ = CommandHelpTemplate
+			}
+
+			HelpPrinter(ctx.App.Writer, templ, c)
+
+			return nil
+		}
+	}
+
+	if ctx.App.CommandNotFound == nil {
+		return Exit(fmt.Sprintf("No help topic for '%v'", command), 3)
+	}
+
+	ctx.App.CommandNotFound(ctx, command)
+	return nil
+}
+
+// ShowSubcommandHelp prints help for the given subcommand
+func ShowSubcommandHelp(c *Context) error {
+	if c == nil {
+		return nil
+	}
+
+	if c.Command != nil {
+		return ShowCommandHelp(c, c.Command.Name)
+	}
+
+	return ShowCommandHelp(c, "")
+}
+
+// ShowVersion prints the version number of the App
+func ShowVersion(c *Context) {
+	VersionPrinter(c)
+}
+
+func printVersion(c *Context) {
+	_, _ = fmt.Fprintf(c.App.Writer, "%v version %v\n", c.App.Name, c.App.Version)
+}
+
+// ShowCompletions prints the lists of commands within a given context
+func ShowCompletions(c *Context) {
+	a := c.App
+	if a != nil && a.BashComplete != nil {
+		a.BashComplete(c)
+	}
+}
+
+// ShowCommandCompletions prints the custom completions for a given command
+func ShowCommandCompletions(ctx *Context, command string) {
+	c := ctx.App.Command(command)
+	if c != nil {
+		if c.BashComplete != nil {
+			c.BashComplete(ctx)
+		} else {
+			DefaultCompleteWithFlags(c)(ctx)
+		}
+	}
+
+}
+
+var appHelpTextFmt = `NAME:
+   %s - %s
+
+USAGE:
+   [global options] command [command options] [arguments...]
+
+VERSION:
+   %s
+
+DESCRIPTION:
+   %s
+
+COMMANDS:
+%s
+GLOBAL OPTIONS:
+%s
+`
+
+var cmdHelpTextFmt = `NAME:
```

Comment:
This text format should conditionally include a sub-commands section.

Reply:
I don't understand, do you mean that in a command help I should list the subcommands somehow? I don't see that in the [original template](https://github.com/urfave/cli/blob/master/template.go#L38).

Also, I just realized that our code [here](https://github.com/mendersoftware/mender/blob/master/cli/cli.go#L118) was redefining the subcommand template, but it will not have any effect anymore so I am removing it.

Reply:
It's a bit confusing, but the SubCommandHelpTemplate is used if the command has any sub-commands: https://github.com/urfave/cli/blob/master/help.go#L190
I'm not sure why they do it like this when they can easily merge the templates, but I would suggest utilizing printf to just conditionally add a command section if any sub-command is present in the struct.

Reply:
Thanks!

I think that upstream uses SubCommandHelpTemplate for actually "commands with subcommands". That terminology seems to be all around help.go

Anyway, this was great because I actually found that with the "minimal" version these "commands with subcommands" were wrongly printed, not only missing the subcommands (as I think you were pointing out) but actually using the "app" format. See [this line](https://github.com/lluiscampos/cli/commit/0d6847ae9ffc664d55a3c8bf0d3395e9fb127c15#diff-c403ca73c89e0b507f69b2d17779ab6eR78-R79) to better understand what I mean.